qingstor/qsctl/utils.py

```diff
@@ -42,6 +42,69 @@
 is_python3 = (_ver[0] == 3)
 
 
+class CompletedSizeRecorder(object):
+    """
+    This class records completed size of downloading object and upload_id of
+    uploading a large object.
+
+    - For downloading:
+
+    After getting one chunk (32M) and writing to local file, use save_record
+    to append current completed size record. When resuming downloading, use
+    get_record to acquire the size of content written to local file in previous
+    download. After downloading one object, remove all records of the object
+    from record file.
+
+    - For uploading:
+
+    Store upload_id for uploading a large object via multipart API.
+
+    - Record format:
+
+    <bucket>/<key>%#&<record_type>#<actual-record>\n
+
+    record_type: RECORD_DOWNLOAD, RECORD_UPLOAD
+    """
+
+    def __init__(self, record_filename):
+        if os.path.exists(record_filename):
+            self.file = open(record_filename, "r+")
+        else:
+            self.file = open(record_filename, "w+")
+        self.separator = "%#&"
```

Comment:
I don't think it a good idea to use this separator, key could be an UTF-8 character, even a key "%#&".